### PR TITLE
common: DB adapter for blockwatch

### LINKED ISSUE #1025: common: Add DB methods adhering to blockwatch.MiniHeaderStore
**Is your feature request related to a problem? Please describe.**

#1024 will define the `MiniHeaderStore` interface which the stack can use to store block headers. We need an implementation of this interface.

**Describe the solution you'd like**

We can use the SQLite DB to store block headers and update `common.DB` with methods so that it adheres to the `MiniHeaderStore` interface.

### PR DESCRIPTION
**Pull request by kyriediculous**
**What does this pull request do? Explain your changes. (required)**
This PR implements the necessary methods for the sqlite DB in the go-livepeer `common` package to adhere to the `MiniHeaderStore` interface in the `blockwatch` package. 

These methods allow us to store the stack that the `blockwatch` package maintains in the DB.

- block numbers are stored as `int64`
- blockhash and parenthash are stored as hex encoded `string` 
- a block's event logs are stored as JSON encoded `string` 

**Specific updates (required)**
- Added a new table `blockheaders` to the schema 
- Added 4 new queries:  `inserMiniHeader`, `findLatestMiniHeader`, `findAllMiniHeadersSortedByNumber` and `deleteMiniHeader`
- Added methods that adhere to the `MiniHeaderStore` interface that implement the new queries
- Added helpers to encode and decode event logs to JSON for storage.
- Added unit tests `db_test.go` to test the new methods

**How did you test each of these updates (required)**
ran unit tests

**Does this pull request close any open issues?**
Fixes #1025 

**Checklist:**
- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: common/db.go
Comment (yondonfu) [inline common/db.go:216]: I guess this is a leftover statement that we forgot to remove after deciding to use `selectOrchs` exclusively. Since this removal is not related to the addition of `MiniHeader` persistence in the DB, it could be a good idea to push this removal into its own distinct commit (we can also remove `filterOrchs` from the `DB` type definition as well).

File: common/db.go
Comment (yondonfu) [inline common/db.go]: "eturns" -> "returns"

File: common/db.go
Comment (yondonfu) [inline common/db.go]: I think the primary key should actually be `hash` since each block header is uniquely identified by its hash (there could be 2 block headers with the same `number` but different hashes).

File: common/db.go
Comment (yondonfu) [inline common/db.go]: If we make `hash` the primary key (see my comment below), then we should add an index for `number` to speed up the `findLatestMiniHeader` and `findAllMiniHeadersSortedByNumber` queries.

File: common/db.go
Comment (yondonfu) [inline common/db.go]: Should be `deleteMiniHeader` to match the style used in the other error messages?

File: common/db.go
Comment (yondonfu) [inline common/db.go:223]: Missing `d.Close()` in the error handlers for all of the statement preparations except for `deleteMiniHeader`

File: common/db.go
Comment (yondonfu) [inline common/db.go]: Need to pass `&logsEnc` into `rows.Scan()` or else it will always be the empty string

File: common/db.go
Comment (yondonfu) [inline common/db.go]: Since `encodeLogsJSON()` returns a `[]byte` and `decodeLogsJSON()` accepts a `[]byte`, I think storing `logs` as a `BLOB` would allow us to skip the extra `string <-> []byte` conversion step.

File: common/db.go
Comment (yondonfu) [inline common/db.go:572]: We should add a `defer rows.Close()` here.

File: common/db.go
Comment (yondonfu) [inline common/db.go]: "Miniheader" -> "MiniHeader"

File: common/db.go
Comment (yondonfu) [inline common/db.go]: I think we could be ok without the length check here since `common.Hash` is just a [type alias](https://github.com/ethereum/go-ethereum/blob/master/common/types.go#L47) for a `[]byte` of length 32.

File: common/db.go
Comment (yondonfu) [inline common/db.go]: I think we could be ok without the length check here since common.Hash is just a type alias for a []byte of length 32.

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: Can kill space?

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: I don't think this function needs to accept `*testing.T` as an argument.

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go:671]: Test the error paths for `InsertMiniHeader()`?

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go:674]: Test the error paths for `FindLatestMiniHeader()`?

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go:756]: Could add a few additional tests:

- When the header doesn't exist
- When there are > 1 headers already in the DB

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go:674]: I think its also worth checking that the `MiniHeader` returned by `FindLatestMiniHeader()` or `FindAllMiniHeadersSortedByNumber()` can contain multiple logs (at the moment looks like we're always inserting `defaultMiniHeader()` which only has a single log).

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: common/db.go
Comment (kyriediculous) [reply common/db.go:216]: How can I extract that into it's seperate commit since it's already part of this commit? thanks !

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Will the queries automatically make use of the index?

Is the efficiency gain here for sorting the blocknumber in descending order on those queries?

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go:674]: This is being checked in the Delete test, where after there are no more remaining entries `FindLatestMiniHeader` returns an error. I have added a comment there for clarity.

I didn't add a seperate test for `FindLatestMiniHeader` because its functionality is being tested in the other tests and it's correct working is crucial for those tests to pass.

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Found an answer

> When you use ORDER BY without an index on the column to be sorted, SQLite needs to build up a temporary data structure that contains the sorted results each time the query is executed. That data structure will use up space in memory (or on disk, depending on the circumstances) and will be torn down after the query is executed. This can be a lot of extra work that might noticeably degrade performance.

[Squeezeing performance from SQLite indexes: Indexes? Indexes!](https://medium.com/@JasonWyatt/squeezing-performance-from-sqlite-indexes-indexes-c4e175f3c346)

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go:674]: Fixed in ced063e by adding a second log for uneven indices in tests and asserting the length of the logs contained in the MiniHeader

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go:756]: Fixed in ced063e

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go:671]: Fixed in ced063e

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: Fixed in ced063e

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go:572]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go:223]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (kyriediculous) [reply common/db.go]: Fixed in ced063e

File: common/db.go
Comment (yondonfu) [reply common/db.go:216]: You can use `git add -p` which triggers git's patch mode (see this [blog post](https://johnkary.net/blog/git-add-p-the-most-powerful-git-feature-youre-not-using-yet/) for a more in-depth treatment of the command).

Suppose 8f4ac86 is the latest HEAD commit and the PR is approved already. You can do the following:

```
# This will unstage all the changes for latest HEAD commit
git reset HEAD~1

# Run git add in patch mode
git add -p
```

If the commit that you care about is further back then you can either use `HEAD~N` where N is the # of commits back from the HEAD or you can reference the commit hash.

The last command will open an interactive console that allows you to select "hunks" (basically chunks of code changes that Git detects) to stage for a commit. You can stage the hunk that corresponds to this change and then add a specific commit for it.

File: common/db.go
Comment (yondonfu) [reply common/db.go]: > Will the queries automatically make use of the index?

Yep.

> Is the efficiency gain here for sorting the blocknumber in descending order on those queries?


Yaeh - adding an index can help avoid the sort because the index can allow the [rows to be delivered in the required order](https://use-the-index-luke.com/sql/sorting-grouping/indexed-order-by).

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: Looks like `err` is compared against itself here when we want to compare it with the expected error string

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: Looks like `err` is compared against itself here when we want to compare it with the expected error string

File: common/db_test.go
Comment (yondonfu) [reply common/db_test.go:674]: Can we add a sanity check that the logs are correctly encoded/decoded as well?

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: This comment is a bit confusing due to the `i%2 == 0` - might be clearer to say that even index = 1 log & odd index = 2 logs

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go:674]: Aren't we checking that at line 676 where we compare that the inserted value equals the retrieved value from the DB? Or are we talking about different things?

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: fixed in 72b6c06

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: fixed in 72b6c06

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: fixed in 72b6c06

**Comment by kyriediculous** [comment]
In 0ca92c0 `FindLatestMiniHeader` no longer returns an error, but nil instead, when it can't find a result in DB.

If we'd return an error `stack.Peek` would error out on startup with an empty stack.

File: common/db_test.go
Comment (yondonfu) [reply common/db_test.go:674]: I wasn't sure if `testify/assert` handled the equality comparison for a nested array of structs properly, but after trying it out locally, it doesn't seem to be a problem. Never mind!

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: Just a heads up - [assert.Equal()](https://godoc.org/github.com/stretchr/testify/assert#Equal) will compare the referenced values when comparing 2 pointers so we don't need to explicitly deference pointers in this test file.

File: common/db_test.go
Comment (yondonfu) [inline common/db_test.go]: Isn't it the opposite? i.e. `i % 2 == 0` (even) -> 1 log, `i % 2 != 0` (odd) -> 2 logs

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: Cool, I thought so but wasn't entirely sure!

File: common/db.go
Comment (j0sh) [inline common/db.go:36]: Would `getMiniHeaders` suffice here as a variable name? That way we aren't re-indenting a large block, and it is more concise while not being *too* specific. Same goes for the function name, `GetMiniHeaders`.

Just curious how reorgs are planned to be handled here?

Comment (kyriediculous) [reply common/db.go:36]: I thought it would be nice for the queries to be named according to the exported method that implements those queries. While we could rename the queries, we can't rename the exported methods as they need to adhere to the `MiniHeaderStore` interface introduced in https://github.com/livepeer/go-livepeer/pull/1030/files

> Just curious how reorgs are planned to be handled here?

Re-orgs are handles by the blockwatch package itself (`blockwatch.buildCanonicalChain()`), which maintains a stack of blocks. in this PR we implement only a persistant storage for that stack. 
If we find a re-org we will pop off blocks off the stack until the point of the re-org and then start Pushing blocks again. 

eg. `Stack.Pop` uses `DB.FindLatestMiniHeader()` and then `DB.DeleteMiniHeader(latestMiniHeader.Hash)` but DB doesn't need to know about any of the reorg logic.

Comment (j0sh) [reply common/db.go:36]: Sounds like the interface could also be renamed if we're just importing everything :) But that's me - just renaming the query handle itself is fine here too.

Thanks for the explanation on reorgs! I would have thought it would build all chains as a tree and use the longest / deepest as the canonical chain, but I guess it makes sense to prune the DB occasionally if a reorg is detected.

Comment (kyriediculous) [reply common/db.go:36]: Yeah we could have, but we decided to move fast and just port blockwatch package over from 0x-mesh and remove dependencies on their codebase. The latter included moving away from a specific database implementation they used to an interface that defines the same methods so we can plug in our own DB. 

0x will make their blockwatch package standalone in the coming weeks which we'll then contribute back to, this is more of an intermediate solution that will behave very similar to the eventual API.  I also had a few optimisations in mind but in today's streamflow design meeting we decided to hold off on those to not diverge from the original API too much and focus in implementing the livepeer specific event watching logic first.

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: Ah yes you're right because we are reversing the order that headers are appended to `added`. 

Fixed in cb0ca5e

File: common/db_test.go
Comment (kyriediculous) [reply common/db_test.go]: Fixed in cb0ca5e

**Comment by yondonfu** [comment]
Changes look good to me.

Re: [this comment thread](https://github.com/livepeer/go-livepeer/pull/1038#discussion_r313003414) - I would be ok keeping the current query name, but I'll leave that up to you.

When you're ready to merge, I recommend trying out `git add -p` to separate the code change mentioned in this [comment](https://github.com/livepeer/go-livepeer/pull/1038#discussion_r312597104) in addition to squashing the fixup commits as it is a useful tool to be familiar with IMO.

**Comment by kyriediculous** [comment]
rebased and created a seperate commit for the cleanup of `filterorchs` references

**Review by yondonfu — APPROVED** [review]
LGTM! 🚢 

I pushed 9523478 to fix an import issue in `eth/blockwatch/client.go`

**Pull request merged**
